src/asl.ts

```diff
@@ -430,19 +430,63 @@ export const isWaitState = (state: State): state is Wait => {
   return state.Type === "Wait";
 };
 
+type EvalExprHandler<Output extends ASLGraph.Output = ASLGraph.Output> = (
+  output: Output,
+  context: EvalExprContext
+) => ASLGraph.NodeResults;
+
 export interface EvalExprContext {
   /**
-   * returns a possibly mutated output variable as json path.
-   *
-   * If the output was a {@link ASLGraph.LiteralValue}, a new state will be added that turns the literal into a json path.
-   * If the output was a {@link ASLGraph.JsonPath}, the output is returned.
+   * Callback provided to inject additional states into the graph.
+   * The state will be joined (@see ASLGraph.joinSubStates ) with the previous and next states in the order received.
    */
-  normalizeOutputToJsonPath: () => ASLGraph.JsonPath;
+  addState: (state: ASLGraph.NodeState | ASLGraph.SubState) => void;
+}
+
+type EvalContextHandler = (context: EvalContextContext) => ASLGraph.NodeResults;
+
+export interface EvalContextContext {
   /**
    * Callback provided to inject additional states into the graph.
    * The state will be joined (@see ASLGraph.joinSubStates ) with the previous and next states in the order received.
    */
   addState: (state: ASLGraph.NodeState | ASLGraph.SubState) => void;
+  /**
+   * Evaluates a single expression and returns the {@link ASLGraph.Output}.
+   *
+   * Any require states will be merged in with the output.
+   *
+   * This method is the same as {@link ASL.evalExpr}, but can be used multiple times within a single {@link ASL.evalContext}.
```

Comment:
it can, but the outputs would need to be merged together.

```ts
return this.evalExpr(expr1, (output1) => {
   const states = this.evalExpr(expr2, (output2) => {
       return ... some states...
   })

   return ASLGraph.joinSubStates(expr,states, ... other states ... );
})
```

vs

```ts
return this.evalContext(expr, ({ evalExpr }) => {
   const output1 = evalExpr(expr1);
   const output2 = evalExpr(expr2);

   ... do something with the outputs ...
})
```

`evalExpr` is a simpler contract that evaluates a single expression, accumulates any states behind the scenes, gives the output to callback, and then merges the accumulated state with the state returned by the callback.

`evalContext` doesn't immediately evaluate an expression, but provides methods that can be used to evaluate any expression as needed. While in the context (callback) any states generated by the helper methods are accumulated and merged at the end with the states returned from the callback.